### ISSUE TITLE
Gemma 4 E4B on Metal: deterministic NaN logits at specific total prompt lengths (safetensors + ISQ q4k/q8_0, still at c22c2e2b)

### ISSUE THREAD

**Issue by dirvine** [EricLBuehler/mistral.rs#2214]
## Summary

`google/gemma-4-E4B-it` (safetensors + ISQ) on **Metal** deterministically produces **NaN logits when a request's *total* prompt length lands in narrow windows of sequence lengths** (windows are ~10 tokens wide and recur periodically). The same payload fails every time, on a fresh process; adding or removing a few dozen characters of prompt text moves it in/out of the failing window.

Likely the same root cause as #2051 (NaN logits / hangs on Gemma 4 "on specific prompts", CUDA) — the length-window mechanism would explain why rephrasing the prompt there made the failure disappear: a different phrasing is a different total length.

Still present at master `c22c2e2b` (2026-06-11). Also present at `ba2f787`.

## Environment

- MacBook Pro, Apple M5 Max, 128 GB — Metal backend
- macOS 26.4 (25E246), rustc 1.96.0
- mistralrs 0.8.3, git rev `c22c2e2b622a815821e59056c2b5952b4cbbe010` (Rust API, not mistralrs-server)
- Model: `google/gemma-4-E4B-it` (safetensors, BF16) with in-situ quantization — reproduces with **both `IsqType::Q4K` and `IsqType::Q8_0`**
- Loaded via the auto `ModelBuilder` (E4B registers as `Gemma4ForConditionalGeneration`), prefix cache disabled (`with_prefix_cache_n(None)`)

## Error

With the default device sampler the failure surfaces mid-stream as:

```
inference failed: invalid Metal top-k softmax normalizer
```

With CPU sampling (`set_sampler_topk(160)`, i.e. top_k > MAX_DEVICE_TOP_K) the true error surfaces instead:

```
inference failed: Invalid sampling probability at index 2: NaN. The model likely produced NaN/Inf logits.
```

Since CPU sampling still hits NaN, the logits themselves are NaN — it's not the Metal top-k kernel, that kernel is just where NaN first explodes with device sampling.

## Minimal deterministic repro

The failing prompt needs no special content — neutral filler triggers it at the right length:

```python
WORDS = ("alpha beta gamma delta epsilon zeta eta theta iota kappa lambda mu "
         "nu xi omicron pi rho sigma tau upsilon phi chi psi omega").split()

def filler(n_chars: int) -> str:
    parts, total, i = [], 0, 0
    while total < n_chars:
        w = WORDS[i % len(WORDS)]
        parts.append(w)
        total += len(w) + 1
        i += 1
    return " ".join(parts)[:n_chars]

system = filler(38_000)   # NaN at q4k AND q8_0 on our machine
user   = filler(633)
```

Driven through the Rust API as:

```rust
let model = ModelBuilder::new("google/gemma-4-E4B-it")
    .with_isq(IsqType::Q4K)        // Q8_0 reproduces too
    .with_prefix_cache_n(None)
    .with_logging()
    .build()
    .await?;

let request = RequestBuilder::new()
    .set_sampler_max_len(32)
    .set_sampler_topk(160)         // CPU sampling → surfaces the true NaN error
    .add_message(TextMessageRole::System, system)
    .add_message(TextMessageRole::User, user);
```

In a sweep of system-prompt sizes 30,000 → 42,000 chars in 400-char steps (user message fixed at 633 chars), every size up to 37,600 passed and **38,000 failed**, on both q4k and q8_0.

## Observations

- **Deterministic per payload**: the same request fails on every attempt, including on a freshly started process.
- **Windows are very narrow**: with a different prompt family (29,730-char system + 36 tool JSON schemas), the exact original length **passed** while **+40 characters failed** — roughly a 10-token-wide window.
- **Windows recur**: padding a failing prompt by ~50–1,600 tokens makes it pass; ~+3,200 tokens lands in the next failing window.
- **Not content- or modality-specific**: plain-text-only requests fail; with/without tool schemas fails; we originally hit this on audio-bearing requests (Gemma 4 audio input) but audio is not required.
- **Not the prefix cache** (disabled throughout).
- q8_0 fails *less often* across arbitrary real-world prompts than q4k in our app telemetry, but at the lengths above it fails identically — higher-precision ISQ shifts/narrows the windows, it does not remove them.

This smells like a prompt-length-dependent edge in the Metal attention/ISQ kernel tiling for this architecture (Metal prefill is single-pass here; chunked prefill is CUDA-only as far as we can tell).

## Workaround we ship

On the NaN signature we retry the request with system-prompt padding of ~50/310/1,040 tokens — one of the paddings reliably exits the window. Ugly but 100% effective so far.

Happy to provide the raw failing request payloads (JSON), the sweep harness, or to run diagnostics/patches on Metal hardware.

**Comment by dirvine** [comment]
Root cause found and fixed upstream in candle: huggingface/candle#3625

The NaN logits at periodic prompt lengths on Metal come from candle's steel
attention prefill kernel (`call_sdpa_full` in
`candle-metal-kernels/src/kernels/sdpa.rs`) being called with **both** an
explicit additive mask **and** `do_causal = true`. The explicit mask already
encodes causality, so `do_causal` double-applies it (the in-shader `kb_lim`
block-truncation plus the causal-triangle loop). At sequence lengths where the
partial K-tile boundary lines up unfavourably, a query row ends up fully masked
to `-inf` → softmax normalizer `sum(exp) == 0` → final divide computes
`0 / 0 = NaN`. It's periodic in total prompt length and quant-independent
(reproduces on Q4K and Q8_0 ISQ).

The fix is one line — skip the redundant in-kernel `do_causal` when an explicit
mask is present:

```rust
let do_causal = do_causal && !has_mask;
```

Verified on Apple M5 Max with Gemma-4-E4B:
- deterministic 38,000-char-prompt repro: FAIL → PASS (Q4K and Q8_0)
- 48-length sweep across 5 periodic windows: 19/48 FAIL → 0/48 (Q4K), 0/48 (Q8_0)
- clean-decode throughput unchanged (~70–74 tok/s)

Once the candle PR lands, bumping the candle pin here should resolve this.

**Comment by EricLBuehler** [comment]
Thank you @dirvine! I merged your PR to candle and will update it here.
